src/ngraph/runtime/reference/batch_norm.hpp

```diff
@@ -85,13 +92,38 @@ namespace ngraph
                         auto channel_beta = arg1[c];
 
                         auto input_index = arg2_transform.index(arg2_coord);
-                        auto normalized = (arg2[input_index] - channel_mean) /
-                                          (std::sqrt(channel_var + eps_casted));
-                        out0[input_index] = normalized * channel_gamma + channel_beta;
+                        out3[input_index] = arg2[input_index] - channel_mean;
+                        out4[input_index] =
+                            out3[input_index] / (std::sqrt(channel_var + eps_casted));
+                        out0[input_index] = out4[input_index] * channel_gamma + channel_beta;
                     }
                 }
             }
 
+            template <typename T>
+            void batch_norm_three_outputs(double eps,
+                                          const T* arg0,
+                                          const T* arg1,
+                                          const T* arg2,
+                                          T* out0,
+                                          T* out1,
+                                          T* out2,
+                                          const Shape& arg2_shape)
+            {
+                std::unique_ptr<T[]> centered(new T[shape_size(arg2_shape)]());
```

Comment:
It's a dynamically created array since the number of elements isn't known at compile-time; the `unique_ptr` ensures it gets deleted when `batch_norm_three_outputs_with_intermediates` returns.

Reply:
@diyessi are we sure it does this in this *particular* case? because the very next thing we do with those `unique_ptr`s is to call `get` on them which makes those `unique_ptr`s to release the ownership over those arrays, so at the point when the unique_ptrs get destructed, they don't own any storage anymore.
I'd suggest using `std::vector<T>` and `data()` to get the underlying storage. Since vectors' lifetime covers intermediate arrays lifetime and yet the vectors would get destroyed after a function is done, this would give us semi-robust memory management w/o needing to call `delete` explicitly

Reply:
Just a drive-by comment. `get` on a `unique_ptr` does not make it release ownership of it's data, it only returns a pointer to it's data. The `unique_ptr` is not destructed by calling `get`